### ISSUE TITLE
Fixing the sbi runner to make sure it can do things other than SNPE-C

### ISSUE THREAD

**Issue by maho3** [maho3/ltu-ili#32]
We've recently made changes to the [SBIRunner](https://github.com/maho3/ltu-ili/blob/main/ili/inference/runner_sbi.py) object to allow it to handle [ensemble posteriors](https://github.com/maho3/ltu-ili/blob/0943be9deba798c9f8d47497e70f108d0f92950f/ili/inference/runner_sbi.py#L178). These ensemble posteriors use a sbi.utils object called [Neural Posterior Ensemble](https://github.com/mackelab/sbi/blob/57e303e5cb02226ad49b07ae29f1fe9f284b5047/sbi/utils/posterior_ensemble.py#L19), which is only supposed to work for [SNLE](https://www.mackelab.org/sbi/reference/#sbi.inference.snle.snle_a.SNLE_A) and [SNPE](https://www.mackelab.org/sbi/reference/#sbi.inference.snpe.snpe_c.SNPE_C) inference engines. We currently hardcode default behavior to use this ensembling protocol, preventing us from using other inference engines such as [SNRE](https://www.mackelab.org/sbi/reference/#sbi.inference.snre.snre_a.SNRE_A).

Also, even though the sbi documentation says it should work for SNLE, I can't get it to run when I change the model parameter in [sample_inference_config.yaml](https://github.com/maho3/ltu-ili/blob/main/examples/configs/sample_inference_config.yaml) to:
```
model:
  module: 'sbi.inference'
  class: 'SNLE_A'
  neural_posteriors:
    - model: 'maf'
    - model: 'mdn'
```
Running this^ config with [example_inference.py](https://github.com/maho3/ltu-ili/blob/main/examples/example_inference.py) I get an error:
```
Traceback (most recent call last):
  File "/Users/matthewho/git/ltu-ili/examples/example_inference.py", line 27, in <module>
    runner()
  File "/Users/matthewho/git/ltu-ili/ili/inference/runner_sbi.py", line 173, in __call__
    density_estimator = model.train(
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/sbi/inference/snle/snle_base.py", line 229, in train
    train_losses = self._loss(theta=theta_batch, x=x_batch)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/sbi/inference/snle/snle_base.py", line 396, in _loss
    return -self._neural_net.log_prob(x, context=theta)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/nflows/distributions/base.py", line 40, in log_prob
    return self._log_prob(inputs, context)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/nflows/flows/base.py", line 38, in _log_prob
    embedded_context = self._embedding_net(context)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1427, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/torch/nn/modules/container.py", line 204, in forward
    input = module(input)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1427, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/homebrew/anaconda3/envs/pt/lib/python3.10/site-packages/sbi/utils/sbiutils.py", line 173, in forward
    return (tensor - self._mean) / self._std
RuntimeError: The size of tensor a (3) must match the size of tensor b (10) at non-singleton dimension 1
```

So basically, the only sbi engine that our code consistently works for is SNPE_C. The goal here is to fix our SBIRunner object's configuration so that it can dynamically parameterize all or most of the [inference engines in the sbi toolkit](https://www.mackelab.org/sbi/reference/#inference). This may require you to remove ensembling for engines such as SNRE, or for you to change ensembling to be at the engine level and not at the posterior model level.

**Comment by maho3** [comment]
Chirag provided [this code](https://github.com/modichirag/HySBI/blob/main/scripts/snle/sample_sweep.py#L7) as an example for efficient emcee samplers of sbi.SNLE posteriors
